cpp/src/arrow/util/async_generator.h

```diff
@@ -1338,11 +1342,34 @@ class BackgroundGenerator {
       return next;
     }
 
+#if (defined(__GNUC__) || defined(__clang__))
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wclass-memaccess"
+#endif
+
+    void SetWorkerThreadId(const std::thread::id tid) {
```

Comment:
Of course, we could also not bother with this complication, since `worker_thread_id` is only used for debugging.

Reply:
True, but the assert catches an otherwise potentially subtle issue in how BackgroundGenerator is meant to be used.  It's not entirely obvious that you aren't allowed to dispose of it from the worker thread and it a fair amount of debugging to discover this root cause.

What was the issue?  Is it because the worker thread id might be changing (from set to unset) when the assert in cleanup happens?

Reply:
>  Is it because the worker thread id might be changing (from set to unset) when the assert in cleanup happens?

Exactly, yes.

Reply:
If this neat trick to turn a `thread::id` into an `int64_t` is indeed reliable it might be nice to add `int64_t arrow::internal::GetCurrentThreadId()`.  Probably in `thread_pool.h/cc`.

Reply:
Will do.

Reply:
See https://github.com/apache/arrow/pull/10644